File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/di/NoNullableInjection.kt
package com.perrystreet.woof.konsist.di

import com.perrystreet.woof.konsist.Assertions.assertFalse
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.KonsistUtils.hasKoinDefinition
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class NoNullableInjection : BehaviorSpec() {
    init {
        Given("A constructor parameter of a Koin-managed class") {
            val parameters = KonsistUtils.productionCode.classes()
                .filter { it.hasKoinDefinition() }
                .flatMap { it.primaryConstructor?.parameters.orEmpty() }

            Then("It is not nullable") {
                parameters.assertFalse(message = Message) { it.type.isNullable }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "Koin-managed classes never inject nullable dependencies.",
            why = "A nullable dependency is a missing definition in disguise. Either the dependency exists in the graph or the design is wrong.",
            howToFix = "Make the dependency non-null and provide it in the DI graph, or split the class.",
            badExample = "class DogsRepository(private val dataSource: IDogsDataSource?)",
            goodExample = "class DogsRepository(private val dataSource: IDogsDataSource)",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/DI/NoNullableInjection.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class NoNullableInjection: QuickSpec {
    override class func spec() {
        Given("An injected dependency of a Swinject-managed class") {
            let classes = WoofHarmonize.productionCode.classes().filter { klass in
                WoofHarmonize.diAnnotations.contains { klass.hasAttribute(named: $0) }
            }

            Then("Its stored constants are not optional") {
                classes.assertTrue(message: message) { klass in
                    klass.variables
                        .filter { $0.isStored && $0.isConstant && !$0.modifiers.contains(.static) }
                        .allSatisfy { !$0.isOptional }
                }
            }

            Then("Its initializer parameters are not optional") {
                classes.assertTrue(message: message) { klass in
                    klass.initializers.flatMap(\.parameters).allSatisfy { $0.typeAnnotation?.isOptional == false }
                }
            }
        }
    }

    private static let message = LintRuleMessage(
        rule: "Swinject-managed classes never inject nullable dependencies.",
        why: "A nullable dependency is a missing registration in disguise. Either the dependency exists in the graph or the design is wrong.",
        howToFix: "Make the dependency non-optional and provide it in the DI graph, or split the class.",
        badExample: "@Single\nfinal class DogsRepository { private let dataSource: DogsDataSourceImplementing? }",
        goodExample: "@Single\nfinal class DogsRepository { private let dataSource: DogsDataSourceImplementing }"
    )
}
